crates/containerd-shim-wasmedge/src/lib.rs

```diff
@@ -1,7 +1,80 @@
-pub mod error;
-pub mod executor;
-pub mod instance;
-pub mod oci_utils;
+use anyhow::Context;
+use containerd_shim_wasm::libcontainer_instance::easy::{
+    context::RuntimeContext, engine::Engine, instance::Instance,
+};
+use libcontainer::workload::ExecutorError;
+use oci_spec::runtime::Spec;
+use std::sync::Arc;
+use wasmedge_sdk::{
+    config::{ConfigBuilder, HostRegistrationConfigOptions},
+    plugin::PluginManager,
+    VmBuilder,
+};
+
+const ENGINE_NAME: &str = "wasmedge";
+
+pub type WasmEdgeInstance = Instance<WasmEdgeEngine>;
+
+pub struct WasmEdgeEngine {
+    vm: Arc<wasmedge_sdk::Vm>,
+}
+
+impl Default for WasmEdgeEngine {
+    fn default() -> Self {
+        PluginManager::load(None).unwrap();
+
+        let host_options = HostRegistrationConfigOptions::default();
+        let host_options = host_options.wasi(true);
+        #[cfg(all(target_os = "linux", feature = "wasi_nn", target_arch = "x86_64"))]
+        let host_options = host_options.wasi_nn(true);
+
+        let config = ConfigBuilder::default()
+            .with_host_registration_config(host_options)
+            .build()
+            .unwrap();
+        let vm = VmBuilder::new().with_config(config).build().unwrap();
+        Self { vm: Arc::new(vm) }
+    }
+}
+
+impl Engine for WasmEdgeEngine {
+    fn name() -> &'static str {
+        ENGINE_NAME
+    }
+
+    fn run(&self, ctx: RuntimeContext, _spec: &Spec) -> anyhow::Result<i32> {
```

Comment:
I like how this new interface makes the shim implementors only need to know about their particular runtime and how to configure it.

Reply:
Do you think we could use a similar API for windows?
i.e., do the normal setup for a windows container, and then let the implementor run the runtime inside the container.